doc/design/secret-provisioning.md

```diff
@@ -0,0 +1,167 @@
+        "HolderURL": <URL to the secret release service>,
+        "Kid": <identifier of the key>,
+        "LocalPath": <Path to the file that will store the secret after its retrieval>,
+        "ClientLibPath": <Path to the client library>
+    },
+    # Secret 2
+    {
+        "Name": <name of the secret>,
+        "AttestationModel": <a number between 1 to 3. See attestation models below>
+        "HolderURL": <URL to the secret release service>,
+        "Kid": <identifier of the key>,
+        "LocalPath": <Path to the file that will store the secret after its retrieval>,
+        "ClientLibPath": <Path to the client library>
+    },
+    ...
+]
+```
+
+Users are responsible for getting the dependent client library into the rootfs consumed
+by Mystikos. They are also responsible for linking with the corresponding server library
+in their services that release the secrets.
+
+## Attestation models
+
+At present, we consider 3 attestation models, from 1 to 3, that can be adopted by the
+client library while it generates attestation evidences, and by the server library
+when it verifies such evidences. Higher number indicates more abstraction in the
+attestation model.
+
+* **model 1**: The client generates a key pair, and an attestation report which
+includes the hash of the public key. The public key (but not the private key)
+and the attestation report are sent to the server. The server will verify the report's
+validity, verify the public key hash, and consider the TEE identity while
+deciding whether to release the secret or not.
+
+* **model 2**: The client generates a self-signed x509 certificate with
+an embedded attestation report which includes the hash of the public key.
+The certificate is sent to the server. The server will verify the
+report's validity, verify the public key hash, and consider the TEE
+identity while deciding whether to release the secret or not.
+
+* **model 3**: The client generates attestation evidences necessary to obtain
+a token from an attestation service, such as MAA. The token is sent to the
+server. The server will verify the validity of the token, and verify various
+claims within the token, including but not limited to public key and TEE
+identity, while deciding whether to release the secret or not.
+
+The reference implementation will use SGX report for attestation purposes. This
+does not limit users from providing their own client/server libraries that
+take advantage of attestation capabilities provided by other TEE platforms such
+as AMD SNP.
+
+## The client library
+
+Mystikos will provide a reference implementation of the client library
+that exposes the following APIs to be called by Mystikos runtime while
+performing secret provisioning:
+
+* `generate_attestation_evidence_with_keys` (for attestation model 1)
+* `generate_x509_cert_with_attestation_extension` (for attestation model 2)
+* `generate_attestation_service_token` (for attestation model 3)
```

Comment:
Does the client lib need to invoke certain Mystikos APIs or access certain special file(s) created by the Mystikos runtime?

Reply:
What does the Mystikos runtime do with the output of the three generate_xyz APIs? Treating the output as anopaque buffer and send the buffer to `HolderURL` with `Kid`?

Reply:
Is the server at `HolderURL` expected to return a opaque blob (wrapped secret)?

Reply:
> Does the client lib need to invoke certain Mystikos APIs or access certain special file(s) created by the Mystikos runtime?

The reference client library will use the extended systcalls to generate quotes/certs. But a user-provided client library doesn't have to.

Reply:
> What does the Mystikos runtime do with the output of the three generate_xyz APIs? Treating the output as anopaque buffer and send the buffer to `HolderURL` with `Kid`?

That's exactly right.

Reply:
> Is the server at `HolderURL` expected to return a opaque blob (wrapped secret)?

Correct. The opaque blob is interpreted and secrets are extracted by the client lib.